plugins/workload-management/src/main/java/org/opensearch/plugin/wlm/service/QueryGroupPersistenceService.java

```diff
@@ -0,0 +1,241 @@
+                inflightCreateQueryGroupRequestCount.decrementAndGet();
+                logger.warn("failed to save QueryGroup object due to error: {}, for source: {}", e.getMessage(), source);
+                listener.onFailure(e);
+            }
+
+            @Override
+            public void clusterStateProcessed(String source, ClusterState oldState, ClusterState newState) {
+                restoreInflightValues(queryGroup.getResourceLimits());
+                inflightCreateQueryGroupRequestCount.decrementAndGet();
+                CreateQueryGroupResponse response = new CreateQueryGroupResponse(queryGroup, RestStatus.OK);
+                listener.onResponse(response);
+            }
+        });
+    }
+
+    /**
+     * Get the allocation value for resourceName for the QueryGroup
+     * @param resourceName - the resourceName we want to get the usage for
+     * @param resourceLimits  - the resource limit from which to get the allocation value for resourceName
+     */
+    private double getResourceLimitValue(String resourceName, final Map<ResourceType, Object> resourceLimits) {
+        for (ResourceType resourceType : resourceLimits.keySet()) {
+            if (resourceType.getName().equals(resourceName)) {
+                return (double) resourceLimits.get(resourceType);
+            }
+        }
+        return 0.0;
+    }
+
+    /**
+     * This method will be executed before we submit the new cluster state
+     * @param queryGroup - the QueryGroup we're currently creating
+     * @param currentClusterState - the cluster state before the update
+     */
+    ClusterState saveQueryGroupInClusterState(final QueryGroup queryGroup, final ClusterState currentClusterState) {
+        final Metadata metadata = currentClusterState.metadata();
+        String groupName = queryGroup.getName();
+        final Map<String, QueryGroup> previousGroups = metadata.queryGroups();
+
+        // check if there's any resource allocation that exceed limit of 1.0
+        String resourceNameWithThresholdExceeded = "";
+        for (ResourceType resourceType : queryGroup.getResourceLimits().keySet()) {
+            String resourceName = resourceType.getName();
+            double existingUsage = calculateExistingUsage(resourceName, previousGroups, groupName);
+            double newGroupUsage = getResourceLimitValue(resourceName, queryGroup.getResourceLimits());
+            inflightResourceLimitValues.computeIfAbsent(resourceName, k -> new DoubleAdder()).add(newGroupUsage);
+            double totalUsage = existingUsage + inflightResourceLimitValues.get(resourceName).doubleValue();
+            if (totalUsage > 1) {
+                resourceNameWithThresholdExceeded = resourceName;
+            }
+        }
+        // check if group count exceed max
+        boolean groupCountExceeded = inflightCreateQueryGroupRequestCount.incrementAndGet() + previousGroups.size() > maxQueryGroupCount;
```

Comment:
Can we do this upfront instead of doing this here? We can probably separate this out to a separate component sitting before we submit a task to cluster state. We can process concurrent createGroup requests and keep track of inflightCount. Deduct inflightCount when queryGroup is persisted in cluster stats via listener. This approach might help in reducing pressure on cluster manager queue where we might end up queuing a lot of createGroup unnecessarily only to be rejected eventually.
@ruai0511 @kaushalmahi12 @jainankitk What do you folks think? Possible?

Reply:
Currently if we don't do it here, the restoreInflightValue will be very messy. But if we decide not to have inflight values anymore, we can certainly do upfront checks

Reply:
I was asking whether we could move this validation check upfront(along with inflight value) before submitting a task to cluster state manager. I guess my previous comment caused confusion here. 
As my concern was around a scenario where a user sends 10k concurrent requests which ends up sitting in cluster manager queue and potentially impacting cluster stability.
But we can park this for later as well and create an issue for now as needs more thinking.

Reply:
After some discussion with @sgup432 and @jainankitk, we decide to have the checks inside the update tasks, and remove inflight checks (since the tasks are serial).

Reply:
I think the number of create operations will be infinitely small in comparison to other cluster state tasks. Hence I don't think it will add any value to unnecessarily introduce this complexity just for this functionality.

Reply:
Seems like ClusterManagerTaskThrottler in general takes care of these kind of stuff where it throttles based on a key which is unique to every component adding tasks to the queue. Considering we are already providing a throttling key, this should be enough and nothing needed apart from that. Can be resolved.